### ISSUE TITLE
MSVC process requires /SUBSYSTEM:CONSOLE.

### ISSUE THREAD

**Issue by GingerMoon** [libbitcoin/libbitcoin#939]
@evoskuil 
I am using  grpc c++, libbitcoin and Qt(Qt VS Tools) .
I met the run time error (unhandled exception) below:
![image](https://user-images.githubusercontent.com/8342412/39406784-dbf27ec0-4bee-11e8-9a77-9ef838820cda.png)

However, if I add the linker option /SUBSYSTEM:CONSOLE, the run time error (the unhandled exception) disappears.

However, it's obviously strange to use this option for a GUI application. After adding this option, when the application is executed, both the console and the GUI are opened.

Would you please tell me how to fix this run time error? Any insight would be very much appreciated!

P.S.
The attachment is the options I am using.
[options.txt](https://github.com/libbitcoin/libbitcoin/files/1958740/options.txt)

Best Regards,
Moon

**Comment by evoskuil** [comment]
There are potentially two distinct considerations here. First, the libbitcoin base library initializes the console (and environment) for MSVC builds so that it can translate console I/O to/from wchar_t to char (as UTF-8). This is creating the console dependency that you are seeing. I'll look into how we can mitigate this for non-console processes. This should really only be enabled by the declaration of BC_USE_LIBBITCOIN_MAIN so we can probably make the console init dependent upon that.

The other consideration is that libbitcoin implements [UTF-8 Everywhere](http://utf8everywhere.org), which is potentially a problem for MSVC users, since all strings passed to/from libbitcoin APIs are presumed to be UTF-8, **not** ANSI/MBCS. This is not something that can be modified, MSVC callers must translated to/from UTF-8 and can do so using the libbitcoin::unicode APIs.

**Comment by evoskuil** [comment]
Do you get the runtime error (without adding SUBSYSTEM:CONSOLE) if you make this change in `unicode.cpp` ?

```cpp
// Use bc::cin in place of std::cin, etc.
std::istream& cin = std::cin;
std::ostream& cout = std::cout;
std::ostream& cerr = std::cerr;
```

**Comment by GingerMoon** [comment]
Hello @evoskuil 
Thank you for looking into this issue.
I tried to use your proposed code change, the error still occurs.
![image](https://user-images.githubusercontent.com/8342412/39457192-a56e894e-4d1e-11e8-8ef1-e3b0f724a0f3.png)

**Comment by evoskuil** [comment]
I don’t see how that code can be executing since there should be no other code than invokes it.

**Comment by GingerMoon** [comment]
@evoskuil 
Sorry for the confusion. After rebuilding the whole solution, the run time doesn't appear.
However, the generated Chinese mnemonics cannot be displayed correctly in Qt. 
I tried to use your advice before "MSVC callers must translated to/from UTF-8 and can do so using the libbitcoin::unicode APIs."
But met the run time unhandled exception.

        word_list my_word_list = create_mnemonic(my_entropy_128, language::zh_Hans); // line 1
	std::string s1 = join(my_word_list); // line 2
	std::wstring s4 = to_utf16(s1); // line 3

Before executing line 3, s1 is:
  | Name | Value | Type
-- | -- | -- | --
▶ | s1 | "农 烃 屏 户 忍 训 远 灭 届 避 骗 息" | std::basic_string<char,std::char_traits<char>,std::allocator<char> >

However, when executing line 3, the unhandled exception is 
Unhandled exception at 0x757F3EF2 in Wallet.exe: Microsoft C++ exception: boost::locale::conv::conversion_error at memory location 0x016FB918.
(c:\cppWorkspace\.nuget\packages\boost.1.64.0.0\lib\native\include\boost\locale\encoding_utf.hpp
utf_to_utf)

Is there anything wrong?

**Comment by GingerMoon** [comment]
Hi @evoskuil 
Thank you so much for still replying me quickly and I know it's so late in your time zone!

I am using Qt VS tools (a vs plugin). I created a Qt VS test solution via this plugin in VS2017 and adding some of the compile&link options of libbitcoins-examples as additional options into the test solution's 
 configuration. I also uploaded the code of the whole solution below for your reference.
[wallet_bk.zip](https://github.com/libbitcoin/libbitcoin/files/1966514/wallet_bk.zip)

I also tried to use the code you shared with me in your latest reply but still face the same run time error.

I also tried to to call to_utf16 in the libbitcoin-examples project  by pasting all of the sample code in your previous reply into the main.cpp and using the following code in unicode.cpp:
std::istream& cin = std::cin;
std::ostream& cout = std::cout;
std::ostream& cerr = std::cerr;
But still met the same run time error.
[unicode.cpp.txt](https://github.com/libbitcoin/libbitcoin/files/1966827/unicode.cpp.txt)
[main.cpp.txt](https://github.com/libbitcoin/libbitcoin/files/1966830/main.cpp.txt)

Best Regards,

**Comment by evoskuil** [comment]
No problem. Try the examples code in the otherwise I modified VS solution (without Qt).

**Comment by evoskuil** [comment]
My solution is the VS2017 one checked in (master, but version3 is likely the same behavior). The only difference is in the examples main() which I posted above. Ran it through debugger with no problems.

**Comment by GingerMoon** [comment]
@evoskuil 
I use Qt's API directly as a workaround.
However, I am still not able to fully understand the API to_utf16.
The picture below is why the run time error occurs:
![image](https://user-images.githubusercontent.com/8342412/39618672-5d518a62-4fb7-11e8-853d-c590f7249097.png)

As you mentioned before, there is no need to set the locale. 
As you know, #define BC_LOCALE_UTF8 "en_US.UTF8", so the following code doesn't trigger any run time error:
    auto my_word_list = create_mnemonic(my_entropy_128, language::en);
    auto s1 = join(my_word_list);
    auto s4 = to_utf16(s1);


In Qt, I need to call the APIs below:
QTextCodec::setCodecForLocale(QTextCodec::codecForName("GBK"))
so that Qt can know what codec to use.
I doubt that maybe in your environment there are some other configuration which hasn't been done in my environment. 
I would appreciate if you could share more insight with me.

Best Regards,

**Comment by GingerMoon** [comment]
Hi @evoskuil 
It seems that s1 is local 8-bit encoding, and local 8-bit encoding is not support for to_utf16.  Am I right?

**Comment by evoskuil** [comment]
Libbitcoin does not have any methods that accept ANSI/MBCS code page encoding (what you are referring to as “local”). Your Qt code should be compiled to use *only* “wide” Windows APIs (ending in W) or fully implement UTF8 Everywhere (which can be a bit of a challenge) as does libbitcoin. Working in ANSI code pages is 1990’s tech at the latest.

**Comment by evoskuil** [comment]
The BC_LOCALE_UTF8 define is just a symbol, it is not what makes UTF8 Everywhere work. But create_mnemonic() does generate utf8, like all libbitcoin strings.

So you need to convert to/from UCS2/UTF16 (wchar_t) in your Qt code from/to UTF8 (chat) for libbitcoin. The libbitcoin Unicode utilities exist to simplify this task.

**Comment by evoskuil** [comment]
Libbitcoin uses UTF8 Everywhere because it eliminates very ugly conditional code in cross compilations. In Windows you must use UTF16 to get Unicode. So use UTF8 and translate to/from UTF16 at the edges (for Windows API calls).

**Comment by evoskuil** [comment]
The code I posted above doesn’t reproduce any problem, so I don’t have a repro to work on.

**Comment by GingerMoon** [comment]
Hi @evoskuil,
Finally I found the workaround why the to_utf16 works in your environment but doesn't work in my environment.

Initially due to the compile error C2001: newline in constant, I saved the dictionary.cpp file as UTF-8, however, in windows, it is actually UTF-8 with BOM.

After reverting the change and set the locale in control panel,  the issue is fixed.
![image](https://user-images.githubusercontent.com/8342412/39749895-6bb18dd2-52e6-11e8-8948-d4b13b84fd14.png)

Thank you very much for your kindness!
